### PR TITLE
Template parser bugs

### PR DESCRIPTION
**Pull request by kinglozzer**
Fixes:

- https://github.com/silverstripe/silverstripe-framework/issues/8742
- https://github.com/silverstripe/silverstripe-framework/issues/8845

### CODE REVIEW

File: src/View/SSTemplateParser.peg [hunk 1]
```diff
@@ -1156,7 +1168,9 @@ class SSTemplateParser extends Parser implements TemplateParser
 
     # This is used to remove template comments
 
-    Comment: "<%--" (!"--%>" /(?s)./)+ "--%>"
+    CommentWithContent: '<%--' ( !"--%>" /(?s)./ )+ '--%>'
+    EmptyComment: '<%----%>'
```
Comment (kinglozzer) [inline src/View/SSTemplateParser.peg]: I tried _a lot_ of different things to get this to work as just one rule, but failed :(

**Comment by sminnee** [comment]
Adding a rule for invalid syntax mystified me a bit, although I see we've done that for blocks as well, so I don't think that necessarily blocks merge.

Even with this rule in place, we'd still have issues when you have `{$` and then a letter, but not something that matches `Lookup`. Perhaps `{$broken[` would trigger this situation as a test-case? This PR is no worse than the status quo in that regard, but should we try and incorporate that one too?

It mystified me that the template parser wasn't throwing an error. I wonder if we can get to the bottom of that? I *think* what might be happening is that the top-level `Template` rule is, in fact, not forced to process the _whole_ template. The malformed tag doesn't match any of `(Comment | Translate | If | Require | CacheBlock | UncachedBlock | OldI18NTag | Include | ClosedBlock |
        OpenBlock | MalformedBlock | Injection | Text)` and so parsing simply stops.

If we could sort that out, then the specific "malformed" rules would be principally about improving the error messages.

The comment change looks fine :-)

**Comment by kinglozzer** [comment]
> Even with this rule in place, we'd still have issues when you have `{$` and then a letter, but not something that matches Lookup. Perhaps `{$broken[` would trigger this situation as a test-case? This PR is no worse than the status quo in that regard, but should we try and incorporate that one too?

`{$broken[` will trigger the parse error I’ve added, as `broken` is a valid injection (it just ignores the trailing square bracket):

> [Emergency] Uncaught SilverStripe\View\SSTemplateParseException: Parse error in template on line 32. Error was: Malformed bracket injection {$broken. Perhaps you have forgotten the closing bracket (})?

Other attempts to break this:
- `{$[` - outputs `{$[` in the template
- `{$123` - outputs `{$123` in the template
- `{$$broken` - outputs `{$` in the template (`$broken` is processed separately as a lookup)

> It mystified me that the template parser wasn't throwing an error. I wonder if we can get to the bottom of that? I think what might be happening is that the top-level Template rule is, in fact, not forced to process the whole template. The malformed tag doesn't match any of (Comment | Translate | If | Require | CacheBlock | UncachedBlock | OldI18NTag | Include | ClosedBlock | OpenBlock | MalformedBlock | Injection | Text) and so parsing simply stops.
> 
>If we could sort that out, then the specific "malformed" rules would be principally about improving the error messages.

I think that sounds like a good idea, I’m not sure how to approach it though. Are you suggesting this could replace existing logic (like the new rule here and/or the mismatched open tags rules) or complement them as a more generic “something went wrong with this statement: `X`” handler?

**Comment by emteknetnz** [comment]
This pull request hasn't had any activity for a while. Are you going to be doing further work on it, or would you prefer to close it now?

**Comment by kinglozzer** [comment]
I think this is ready. It’s not perfect, but it’s an improvement on the status quo

**Review by dhensby — APPROVED** [review]

**Comment by dhensby** [comment]
As @sminnee has said, this is an improvement and the bit's that could be improved further are not from a regression.

**Comment by sminnee** [comment]
I think we can merge into 4.6 or 4

**Comment by dhensby** [comment]
Rebased and green 💚

**Review by sminnee — APPROVED** [review]

**Comment by kinglozzer** [comment]
Thanks for sorting Dan

**Pull request merged**